backtracking/generate_parentheses.cpp

```diff
@@ -0,0 +1,97 @@
+/**
+ * @file
+ * @brief Generates all combinations of well-formed parentheses.
+ *
+ * @details a sequence of parentheses is well-formed if each opening parentheses
+ has a corresponding closing parenthesis
+ *  and the closing parentheses are correctly ordered
+ *
+ * @author [Giuseppe Coco](https://github.com/WoWS17)
+
+ */
+
+#include <cassert>
+#include <iostream>
+#include <vector>
+
+class generate_parentheses {
+ private:
+    std::vector<std::string> res;  // Contains all possible valid patterns
+
+    /**
+     * @brief function that implements backtracking
+     *
+     * @param str string build during backtracking
+     * @param n number of pairs of parentheses
+     * @param closed number of closed parentheses
+     * @param open number of open parentheses
+     */
+
+    void makeStrings(std::string str, int n, int closed, int open) {
```

Comment:
It is preferred to declare the functions _outside_ the class and have the prototypes inside only.